### DISCUSSION TITLE
Does Virtua Racing for Genesis work on Sanni?

### DISCUSSION THREAD

**Discussion by PsychoFox11** [sanni/cartreader#768] [Need help]
I've been trying to read my Virtua Racing cart, which I know is a little different with the extra cpu and all. It just detects it as 1mbit and FFFF and dumping it doesn't get good results. I've dumped a lot of carts and have learned how clean they need to be etc. Wondering if my cart has a problem or if this one doesn't work on Sanni. Using HW V5 and 12.4.

**Comment by PsyK0p4T** [comment]
I haven't tried to dump this one yet. @PsychoFox11 can you post the logs generated when you dumped the cart please ?

**Reply by PsychoFox11** [reply]
Yeah I'll try to get that tomorrow, I didn't save the previous attempt. I believe it just saved into a number folder like MD/798 rathar than MD/GameTitle/798.
I'll try again and share the logs here.

**Reply by PsyK0p4T** [reply]
This can happen if the game doesn't have an internal name in header. Empty string is used... but the rom usually get renamed fine if the CRC is found in txt file. Your log should give us more info about the problem here.

**Reply by PsychoFox11** [reply]
Sorry for the delay. Here's the log (bin file written is 128k). Fresh attempt from just now, last was on 12.3 but I remember the same numbers etc. which I'm sure is expected:
```
OSCR HW5 V12.4

[+] Mega Drive/Genesis

[+] Game Cartridge


Cart Info
 
Name: 
bramCheck: FFFF
Size: 1 MBit
ChkS: FFFF
Sram: None
 
Press Button...

[+] Read Rom


Saving to MD/ROM//424/...
[*******************]
Internal checksum...Error
E942

CRC32... 1C9668A4 -> Not found
```

**Reply by PsyK0p4T** [reply]
Quickly looked at the USA rom, and its header looks "normal". I mean, the game has an internal title (no blank string), and a Chks "95AB". I guess you cleaned it already ? Does the cart run fine in a genesis console ?
That cart is so particular, with embedded processor, it could be unsupported :/

**Reply by PsychoFox11** [reply]
To be honest, I need to test it to ensure it still works. I just moved and systems are still packed. I've been using the cart dumper as a way to pass time while I'm working my way toward setting up the game room. Maybe I can dig out one of my Genesis consoles this week and see. It DID work, but probably been years since I played it. I will verify and report as soon as I can!

**Reply by PsyK0p4T** [reply]
Confirmed here too with the european version. The rom is not detected well and can't be dumped with current code. Hardware is quite different than usual MD carts...

```
OSCR HW5 V12.5

[+] Mega Drive/Genesis

[+] Game Cartridge


Cart Info
 
Name: FF
bramCheck: 0046
Size: 1 MBit
ChkS: FFFF
Sram: None
 
Press Button...

[+] Read Rom


Saving to MD/ROM/FF/317/...
[*******************]
Internal checksum...Error
140C
```

![image](https://user-images.githubusercontent.com/87064902/235305788-4020434c-d0b7-4a6b-885e-6a213507acc3.png)

**Reply by Prominos** [reply]
I ran into the same problem and with a bit of minor reverse engineering and trial and error I was able to make this work.

![analysis](https://github.com/sanni/cartreader/assets/133436648/46adfdc0-e2b3-4340-90cc-adf5e1295e54)

I started by tracing the connector pins to their internal components. My first realization being that the rom chip (upper right IC2) is not directly connected to the console address and data bus. It's addresses and data lines are only connected to the Sega Virtua Processor (SVP) which is the big chip in the middle (IC1) that handles drawing polygons (and memory mapping in this case!)

This explains why it was impossible to read the ROM header. The SVP is acting as a middle-man (chip?) between the cartridge connector and the ROM chip, therefore, to access the ROM you need to interface with the SVP.

Second thing I realized is that some pins on the connector that are normally left unconnected in most games are routed in this case.
This includes:

1.  `HSYNC` or `Horizontal Synchronization` : The VDP (Video Display Processor) bring this line low during the Horizontal blank period (in between horizontal line). I am guessing the SVP uses this a bit like an interrupt line would work on a CPU it can do some processing while the screen is not actively being drawn.
2. `AS` or `Address Strobe`: Is brought low to indicate that the address on the bus is valid and not changing.
3. `VCLK` or `M68K Clock` this is the clock signal output by the Mega Drive/Genesis main CPU (The Motorolla 68000) @7.6MHz. If this chip works a bit like a co-processor then it is highly possible that it uses synchronous logic, in that case it needs a clock signal to be able to do anything, internal state changes, inputs and outputs are calculated at clock ticks or edges. I'm pretty sure this pin is also use to interface with Serial EEPROMs (save chips).
4. `ACK` or `Acknowledge`: This is a signal that is output from the SVP to the CPU probably to let it know that it acknowledge the receipt of a special command.
5. `ASEL` (Address select?): I am not sure what this one is exactly but it is explained as `Read or Write on $000000-$7FFFFF region` when low. Since this memory region falls into the cartridge ROM range, I believe it should be low.
6. `WE` or `Write Enable`: Used with save chips like SRAM or EEPROMs to write data to it. In this case my guess is that it is either to "write" a command to the SVP or write data to the DRAM chip in the upper left corner (IC3) (I think the former makes more sense)

With that in mind I checked the MD code release V12.4 and found out that reading from ROM is done by writing the address to the bus, bringing ~OE (Output Enable) low to connect the data bus to the ROM data lines and CEO (Chip Enable) to Enable the ROM chip.

This works on 99% of cases but not in that one. So what I did was to also bring `~ASEL` and `~AS` low along with `~CEO` and `~CAS`. 

The other thing to do was to give a valid clock signal to the SVP, fortunately we have the Adafruit Clockgen. Unfortunately it isn't wired for this use case... I happen to have some individual cartridge slots adapter on hand so I chose the easy solution of soldering a jumper cable between `CLK0` of the clock generator to the `VCLK` pin of the MD connector. It looks a bit like that:

![mod](https://github.com/sanni/cartreader/assets/133436648/5867c339-0e67-4ef4-869a-3859fdbfe055)

Now since pin `VCLK` is already wired to the arduino digital pin 16 (D16) we must make sure that the arduino doesn't interfere with the clock generator so bit 1 of PORT H (D16) is set to input in `DDRH` (Can't remember if it can be set to High-Z instead? Does the arduino have tristated I/O?). This is not compatible with the case when this pin is used to write to writable chip, so the software should be context sensitive to setup the pin and clock generator properly. 

The last step is to initialize the clock generator to output a 7.6MHz signal on `CLK0`, this emulates the MD CPU clock. I basically copy pasted most of the initialization logic from the SNES module (SA-1). It is possible that this could work at lower clock speed or with no clock at all but it needs to be tested more thoroughly as I have only done limited testing. (I was successful dumping the cart using this method multiple times though)

That's it!

Result:

![image](https://github.com/sanni/cartreader/assets/133436648/d0e9a069-3401-4bf9-be3a-727228fa57c8)

Hope this is useful to everyone! It certainly was useful to me! :)

**Reply by RascalDungeon** [reply]
Thanks for this! I have finally been able to dump my copy of Virtua Racing. Worked great!

**Reply by PsyK0p4T** [reply]
Which cart do you have ? We tested the script so far with European and Japanese versions of the game. No feedback for the USA one until now.

**Reply by RascalDungeon** [reply]
USA version. Tried dumping it when I dumped the rest of my collection awhile ago and it was the only cart that didn't work. Saw this post, updated my firmware, and tried again. Now my whole collection is dumped.

**Reply by PsyK0p4T** [reply]
Oh, awesome then. All regions are confirmed now :)
Thanks for feedback (and don't hesitate to submit your redumps to no-intro forums)

**Reply by retrovik** [reply]
Just wanted to say great work on this! I did not realize Virtua Racing was now supported. I updated my HW3 revision with the 12.8 Beta and it worked perfectly for my USA release. That was the last cart in my collection I could not dump. Fantastic!

**Comment by Prominos** [comment]
PS: Ignore the second jumper cable. It was to bring `~HSYNC` low but I tried without it and it works just fine.

**Comment by PsyK0p4T** [comment]
Amazing work, really ! And thanks a lot for all documentations, it will be useful to me too for sure.

Can you share your modified MD.ino ?

With your CLK hardware mod in place + v12.4 original code, have you tried to redump different 'classic' MD/Genesis carts ? Just to confirm that the mod doesn't affect all other usual carts

**Reply by sanni** [reply]
Could be worth a try to pulse pin D16 in the readWord_MD function instead of connecting it to the clock generator similar to how we do it with [eeprom](https://github.com/sanni/cartreader/blob/master/Cart_Reader/MD.ino#L2147). Maybe it's enough for this cart as well.

**Reply by Prominos** [reply]
I have experimented with your idea Sanni and here are my findings:

- Pulsing the clock only once before and after reading: Does not work.
- Pulsing the clock 100 times before and after reading: Works but is really slow to dump.
- Pulsing the clock 5 times before and after reading: Gives the correct game title "Virtua Racing" when reading header data but ultimately failed. (unstable?)
- Pulsing the clock 10 times before and after reading: Works and dumps at a reasonable speed.

Some other thoughts:

- Maybe pulsing after the read is not needed.
- Maybe this could be done only for virtua racing but I am not savvy enough with the code to detect this game in particular.

[MD.txt](https://github.com/sanni/cartreader/files/11586971/MD.txt)

**Reply by PsyK0p4T** [reply]
I've just tested your modifications on the dev branch, and can confirm that European version of Virtua Racing got dumped fine aswell, without any hardware modification :) 
Will try to do more tests tonight, and maybe add conditions. Very nice job !

**Comment by Prominos** [comment]
Here's my ino file (changed the extension so github would accept it). 

[MD.txt](https://github.com/sanni/cartreader/files/11470099/MD.txt)

I reuploaded the original V12.4 to my OSCR, kept the HW mod tried to dump Sonic 2, Sonic 3 and Super Street Fighter II, no issues. SFII gave me an invalid internal checksum but the CRC32 is the same as no-intro and the file was renamed.

sanni, you are probably right with pulsing the clock, I am not sure how many times you would need to pulse it though. I also remember that the 2560 can output a clock signal using its internal oscillator but I can't remember if it can be output to any pins or just specific pins.
